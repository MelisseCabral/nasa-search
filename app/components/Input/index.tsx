import classNames from 'classnames';
import { forwardRef, ForwardRefRenderFunction, HTMLAttributes, PropsWithChildren } from 'react';
import './styles.css';

export type TextTypes = 'text';

interface IInput extends HTMLAttributes<HTMLInputElement> {
  name?: string
  error?: string;
  type: TextTypes;
  value?: string;
  required?: boolean;
}

const Input: ForwardRefRenderFunction<HTMLInputElement, PropsWithChildren<IInput>> = (
  { error, className, ...rest },
  ref
): JSX.Element => {
  return (
    <input
      className={classNames('input-primary', className, error ? 'error-input' : '')}
      {...rest}
      ref={ref}
    />
  );
};

export default forwardRef(Input);