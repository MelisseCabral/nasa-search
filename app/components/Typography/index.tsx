import classNames from 'classnames';
import { FC, HTMLProps, ReactNode } from 'react';
import './styles.css';

type TypographyProps = {
  /**
   * Set the text-align on the component.
   * @default 'inherit'
   */
  align?: 'inherit' | 'left' | 'center' | 'right' | 'justify';
  /*
   * The CSS class name of the wrapper element.
   */
  className?: string;
  /**
   * The content of the component.
   */
  children?: ReactNode;
  /**
   * If `true`, the text will have no margin.
   * @default true
   */
  noGutter?: boolean;
  /**
   * If `true`, the text will have a bottom margin.
   * @default false
   */
  gutterBottom?: boolean;
  /**
   * If `true`, the text will not wrap, but instead will truncate with a text overflow ellipsis.
   *
   * Note that text overflow can only happen with block or inline-block level elements
   * (the element needs to have a width in order to overflow).
   * @default false
   */
  noWrap?: boolean;
  /**
   * If `true`, the element will be a paragraph element.
   * @default false
   */
  paragraph?: boolean;
  /**
   * Applies the theme typography styles.
   * @default 'regular1'
   */
  variant?:
    | 'bold1'
    | 'bold2'
    | 'bold3'
    | 'bold4'
    | 'bold5'
    | 'medium1'
    | 'medium2'
    | 'medium3'
    | 'medium4'
    | 'medium5'
    | 'regular1'
    | 'regular2'
    | 'regular3'
    | 'regular4'
    | 'regular5'
    | 'light1'
    | 'light2'
    | 'light3'
    | 'light4'
    | 'light5'
    | 'inherit';
} & HTMLProps<HTMLParagraphElement>;

const Typography: FC<TypographyProps> = ({
  variant = 'regular1',
  noGutter = false,
  gutterBottom = false,
  align = 'inherit',
  noWrap = false,
  children,
  className = '',
  ...restProps
}) => {
  const rootClassName = classNames(
    'typography',
    variant.includes('title') ? 'title' : '',
    `${variant}-typography`,
    noGutter ? 'noGutter' : '',
    gutterBottom ? 'gutterBottom' : '',
    noWrap ? 'noWrap' : '',
    align ? `text-${align}` : '',
    className ? className : ''
  );

  return (
    <span className={rootClassName} {...restProps}>
      {children}
    </span>
  );
};

export default Typography;