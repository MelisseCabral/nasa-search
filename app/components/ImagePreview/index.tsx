import Image from "next/image";
import Link from "next/link";
import Typography from "../Typography";
import InfoImage from "@/app/interfaces/InfoImage";

interface ImageNasaPreview {
    thumbnailUrl: string;
    data: InfoImage
}

export default function ImagePreview({ thumbnailUrl, data } : ImageNasaPreview) {
  return (
    <div className="py-4 grid grid-cols-3">
      <Link as={`/photo/${data.nasa_id}`} href="/photo/[id]">
        <Image alt="" width={250} height={125} src={thumbnailUrl} className="rounded-md"/>
        <div className="nasaId text-white flex flex-col">
          <Typography variant="regular5" className="pt-2">{data.title}</Typography>
          <Typography variant="light5" className="text-gray-300">{data.location}</Typography>
          <Typography variant="light5" className="text-gray-300">By: {data.photographer}</Typography>
        </div>
      </Link>
    </div>
  );
}