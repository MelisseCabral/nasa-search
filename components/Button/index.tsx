import React, { HTMLAttributes, PropsWithChildren, ReactNode } from 'react';
import './styles.css';
import Typography from '../Typography';

interface ButtonProps extends PropsWithChildren<HTMLAttributes<HTMLButtonElement>>{
  /**
   * Is this the principal call to action on the page?
   */
  primary?: boolean;
  /**
   * How large should the button be?
   */
  size?: 'small' | 'medium' | 'large';
  /**
   * Button contents
   */
  label: string;
  /**
   * Optional click handler
   */
  startAddornment?: ReactNode;
  endAddornment?: ReactNode;
  children?: ReactNode;
  onClick?: () => void;
  type?: "button" | "submit";
}

/**
 * Primary UI component for user interaction
 */
const Button = ({
  primary = false,
  size = 'medium',
  label,
  startAddornment = null,
  endAddornment = null,
  children,
  ...props
}: ButtonProps) => {
  const mode = primary ? 'storybook-button--primary' : 'storybook-button--secondary';
  return (
    <button
      className={['storybook-button', `storybook-button--${size}`, mode].join(' ')}
      {...props}
    >
      {startAddornment}
      <Typography variant={'regular5'} >
        {label}
        {children}
      </Typography>
      {endAddornment}
    </button>
  );
} 

export default Button;