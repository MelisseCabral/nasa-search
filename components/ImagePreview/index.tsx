import Image from "next/image";
import Link from "next/link";
import Typography from "../Typography";
import InfoImage from "@/interfaces/InfoImage";

import './styles.css'

interface ImageNasaPreview {
    thumbnailUrl: string;
    data: InfoImage
}

export default function ImagePreview({ thumbnailUrl, data } : ImageNasaPreview) {
  return (
    <div className="py-4">
      <Link as={`/photo/${data.nasa_id}`} href="/photo/[id]">
        <div className="image-container">
          <Image alt="" layout='fill' objectFit="cover" src={thumbnailUrl} className="rounded-md"/>
        </div>
        
        <div className="nasaId text-white flex flex-col">
          <Typography variant="regular5" className="pt-2">{data.title}</Typography>
          <Typography variant="light5" className="text-gray-300">{data.location}</Typography>
          <Typography variant="light5" className="text-gray-300">By: {data.photographer}</Typography>
        </div>
      </Link>
    </div>
  );
}