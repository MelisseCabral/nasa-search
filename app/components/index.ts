import ImagePreview from './ImagePreview/index';

export { default as Typography } from './Typography';
export { default as Input } from './Input';
export { default as Button } from './Button';
export { default as ImagePreview } from './ImagePreview';