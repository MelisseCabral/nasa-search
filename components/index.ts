import ImagePreview from './ImagePreview/index';
import Modal from './FileDetail/index';

export { default as Typography } from './Typography';
export { default as Input } from './Input';
export { default as Button } from './Button';
export { default as ImagePreview } from './ImagePreview';
export { default as FileDetail } from './FileDetail';