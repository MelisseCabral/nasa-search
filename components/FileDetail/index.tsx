import { Dialog, Transition } from '@headlessui/react';
import React, { HTMLAttributes, PropsWithChildren, useState } from 'react';
import Typography from '../Typography';
import { ImageSelected } from '@/interfaces/ImageSelected';
import Button from '../Button';
import Image from 'next/image'
import DownloadIcon from './../../public/download-icon.svg'
import CloseIcon from './../../public/close-icon.svg'

interface IFileDetail extends PropsWithChildren<HTMLAttributes<HTMLButtonElement>> {
  selected: ImageSelected | undefined
  isOpen: boolean;
  onClose: () => void;
}

const FileDetail: React.FC<IFileDetail> = ({ isOpen, selected, onClose }) => {

  const date = selected ? new Date(selected?.data?.date_created?.toString() ?? '') : null;

  const formattedDate = date instanceof Date ? date.toLocaleDateString() : '';


  async function downloadImage() {
    
    if(selected) {
      const response = await fetch(selected?.thumbnails);
    
      const blobImage = await response.blob();
    
      const href = URL.createObjectURL(blobImage);
    
      const anchorElement = document.createElement('a');
      anchorElement.href = href;
      anchorElement.download = `${selected?.data?.nasa_id}.png`;
    
      document.body.appendChild(anchorElement);
      anchorElement.click();
    
      document.body.removeChild(anchorElement);
      window.URL.revokeObjectURL(href);
    }
  }

  return (
    <Dialog
      as="div"
      className="fixed inset-0 z-50 overflow-y-auto"
      open={isOpen}
      onClose={onClose}
    >
      <div className="flex items-center justify-center h-screen w-screen max-h-screen overflow-auto shadow-xl  bg-indigo">
        {selected?.data && (
          <div className=" text-white h-screen w-screen">
            <div className="w-full border-b border-indigo">
              <div className="justify-between flex w-full ">
                <div className="flex flex-col pl-4 py-4">
                  <Typography variant="regular3">{selected.data.title}</Typography>
                  <Typography variant="light5" className="text-gray-300">By: {selected.data.photographer || 'Unknown'} | Location: {selected.data.location} </Typography>
                </div>
                <div className='flex flex-row justify-center items-center'>
                  <Button onClick={downloadImage} 
                    startAddornment={<DownloadIcon />} 
                    className="text-pink-800 font-regular px-4 justify-center gap-4 bg-pink-300 py-2 mr-3 rounded-md hidden md:flex md:flex-row" 
                    label={''}>
                    Download
                  </Button>

                  <Button onClick={onClose} 
                  startAddornment={<CloseIcon className='m-auto' />}
                  className="bg-transparent text-white px-6 justify-center items-center border-l border-indigo h-[100%] w-20" 
                  label={''}/>
                </div>
              </div>
            </div>
            <div className='py-8'>
              <div className="relative h-[60vh] w-[70vw] m-auto mt-4">
                <Image alt="" layout='fill' objectFit="cover" src={selected.thumbnails} className="rounded-md"/>
              </div>
              <div className='w-[70vw] m-auto border-b border-indigo py-4'>
                <Typography variant='regular3'>Information</Typography>
              </div>
              <div className='w-[70vw] m-auto border-b border-indigo justify-between flex py-4'>
                <Typography variant='regular5' className='text-gray-300'>Created By</Typography>
                <Typography variant='regular5' >{selected.data.photographer}</Typography>
              </div>
              <div className='w-[70vw] m-auto border-b border-indigo justify-between flex py-4'>
                <Typography variant='regular5' className='text-gray-300'>Date created</Typography>
                <Typography variant='regular5' >{formattedDate || 'Unknown'}</Typography>
              </div>
              <div className='w-[70vw] m-auto flex flex-col  py-4'>
                <Typography variant='light4' className='py-4'>Description</Typography>
                <Typography variant='light5'>{selected.data.description}</Typography>
              </div>

              <div className='w-[70vw] m-auto flex flex-row justify-between py-2 border-t border-b border-indigo'>
                <Typography variant='light4' className='py-4'>Keywords</Typography>
                <Typography variant='light5'>{selected.data.keywords.join(", ")}</Typography>
              </div>
            </div>
          </div>
        )}
      </div>
    </Dialog>
  );
};

export default FileDetail;