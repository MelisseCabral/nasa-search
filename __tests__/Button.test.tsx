import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import {Button} from './../components';

describe('Button component', () => {

    test('renders snapshot unchanged', () => {
        const label = 'Click me';
        const { container } = render(<Button label={label}/>)
        expect(container).toMatchSnapshot()
    })
    test('renders button with label', () => {
        const label = 'Click me';
        const { getByText } = render(<Button label={label} />);
        const buttonElement = getByText(label);
        expect(buttonElement).toBeInTheDocument();
    });

    test('calls onClick handler when clicked', () => {
        const onClick = jest.fn();
        const { getByText } = render(<Button label="Click me" onClick={onClick} />);
        const buttonElement = getByText('Click me');
        fireEvent.click(buttonElement);
        expect(onClick).toHaveBeenCalledTimes(1);
    });

    test('applies the "primary" class when primary prop is true', () => {
        const { container } = render(<Button label="Button" primary />);
        expect(container.firstChild).toHaveClass('storybook-button--primary');
    });

    test('applies the correct size class based on the size prop', () => {
        const { container } = render(<Button label="Button" size="large" />);
        expect(container.firstChild).toHaveClass('storybook-button--large');
    });
});
