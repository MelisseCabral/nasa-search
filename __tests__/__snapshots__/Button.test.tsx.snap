// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component renders snapshot unchanged 1`] = `
<div>
  <button
    class="storybook-button storybook-button--medium storybook-button--secondary"
  >
    <span
      class="typography regular5-typography text-inherit"
    >
      Click me
    </span>
  </button>
</div>
`;
